radix-transactions/src/validation/transaction_validator.rs

```diff
@@ -1298,6 +1274,322 @@ impl HeaderAggregation {
     }
 }
 
+pub struct AllPendingSignatureValidations<'a> {
+    transaction_version: TransactionVersion,
+    config: &'a TransactionValidationConfig,
+    root: (
+        PendingIntentSignatureValidations,
+        TransactionValidationErrorLocation,
+    ),
+    non_roots: Vec<(
+        PendingIntentSignatureValidations,
+        TransactionValidationErrorLocation,
+    )>,
+    total_signature_validations: usize,
+}
+
+pub struct SignatureValidationSummary {
+    root_signer_keys: IndexSet<PublicKey>,
+    non_root_signer_keys: Vec<IndexSet<PublicKey>>,
+    total_signature_validations: usize,
+}
+
+impl<'a> AllPendingSignatureValidations<'a> {
+    fn new_with_root(
+        transaction_version: TransactionVersion,
+        config: &'a TransactionValidationConfig,
+        root_intent_hash: IntentHash,
+        signatures: PendingIntentSignatureValidations,
+    ) -> Result<Self, TransactionValidationError> {
+        let intent_signature_validations = signatures.intent_signature_validations();
+        let error_location = TransactionValidationErrorLocation::for_root(root_intent_hash);
+        if intent_signature_validations > config.max_signer_signatures_per_intent {
+            return Err(TransactionValidationError::SignatureValidationError(
+                error_location,
+                SignatureValidationError::TooManySignatures {
+                    total: intent_signature_validations,
+                    limit: config.max_signer_signatures_per_intent,
+                },
+            ));
+        }
+        let notary_signature_validations = signatures.notary_signature_validations();
+
+        Ok(Self {
+            transaction_version,
+            config,
+            root: (signatures, error_location),
+            non_roots: Default::default(),
+            total_signature_validations: intent_signature_validations
+                + notary_signature_validations,
+        })
+    }
+
+    pub fn add_non_root_subintents_v2(
+        &mut self,
+        non_root_subintents: &PreparedNonRootSubintentsV2,
+        signatures: &PreparedNonRootSubintentSignaturesV2,
+    ) -> Result<(), TransactionValidationError> {
+        let non_root_subintents = &non_root_subintents.subintents;
+        let non_root_subintent_signatures = &signatures.by_subintent;
+        if non_root_subintents.len() != non_root_subintent_signatures.len() {
+            return Err(
+                SignatureValidationError::IncorrectNumberOfSubintentSignatureBatches
+                    .located(TransactionValidationErrorLocation::AcrossTransaction),
+            );
+        }
+        for (index, (subintent, signatures)) in non_root_subintents
+            .iter()
+            .zip(non_root_subintent_signatures)
+            .enumerate()
+        {
```

Comment:
We're not validating the length of `signatures`?

Reply:
If there aren't explicit tests for subintent signatures, I'll add them 👍

Reply:
I vaguely remember that I added such a test.

Reply:
We missed a test on signature count on the root manifest, so I've added that.